src/ucp/core/ucp_am.c

```diff
@@ -626,55 +633,74 @@ static ucs_status_t ucp_am_long_first_handler(void *am_arg, void *am_data,
     size_t left;
-    ucp_recv_desc_t *rdesc;
     ucp_ep_h reply_ep;
-    ucp_am_unfinished_t *unfinished;
     ucp_am_mid_hdr_t *m_hdr;
 
-    reply_ep = (f_hdr->super.super.flags & UCP_AM_SEND_REPLY) ?  ep : NULL;
-    left     = f_hdr->total_size - (am_length - sizeof(*f_hdr));
+    left = f_hdr->total_size - (am_length - sizeof(*f_hdr));
 
     if (left == 0) {
         /* Can be a single fragment if send was issued on stub ep */
+        reply_ep = (f_hdr->super.super.flags & UCP_AM_SEND_REPLY) ?  ep : NULL;
         return ucp_am_handler_common(worker, f_hdr + 1, sizeof(*f_hdr),
                                      am_length, reply_ep, am_id, am_flags);
     }
 
-    /* If there are multiple messages, we first check if any of the other
-     * messages have arrived */
-    unfinished = ucp_am_find_unfinished(worker, ep_ext, f_hdr->msg_id);
-    if (unfinished == NULL) {
-        unfinished = ucp_am_alloc_unfinished(worker, ep_ext, f_hdr->msg_id);
-        if (ucs_unlikely(unfinished == NULL)) {
-            return UCS_OK; /* release UCT desc */
-        }
-    }
+    /* This is the first fragment, other fragments (if arrived) should be on
+     * ep_ext->am.mid_rdesc_q queue */
+    ucs_assert(NULL == ucp_am_find_first_rdesc(worker, ep_ext, f_hdr->msg_id));
 
-    /* Alloc buffer for the data and its desc, as we know total_size */
-    unfinished->all_data = ucs_malloc(f_hdr->total_size + sizeof(ucp_recv_desc_t),
-                                     "ucp recv desc for long AM");
-    if (ucs_unlikely(unfinished->all_data == NULL)) {
+    /* Alloc buffer for the data and its desc, as we know total_size.
+     * Need to allocate a separate rdesc which would be in one contigious chunk
+     * with data buffer. */
+    data_rdesc = ucs_malloc(f_hdr->total_size + sizeof(ucp_recv_desc_t),
+                            "ucp recv desc for long AM");
+    if (ucs_unlikely(data_rdesc == NULL)) {
         ucs_error("failed to allocate buffer for assembling UCP AM (id %u)",
                   am_id);
-        return UCS_OK; /* release UCT desc */
+        goto out;
     }
-    unfinished->all_data->flags = UCP_RECV_DESC_FLAG_MALLOC;
-    unfinished->left            = f_hdr->total_size;
-    unfinished->am_id           = am_id;
-    unfinished->total_size      = f_hdr->total_size;
-    unfinished->reply_ep        = reply_ep;
 
-    ucs_queue_for_each_extract(rdesc, &unfinished->mid_rdesc_q, am_queue, 1) {
-        m_hdr = (ucp_am_mid_hdr_t*)(rdesc + 1);
-        memcpy(UCS_PTR_BYTE_OFFSET(unfinished->all_data + 1, m_hdr->offset),
-               m_hdr + 1, rdesc->length - sizeof(*m_hdr));
-        unfinished->left -= rdesc->length - sizeof(*m_hdr);
-        ucp_recv_desc_release(rdesc);
+    data_rdesc->flags = UCP_RECV_DESC_FLAG_MALLOC;
+    left              = f_hdr->total_size;
+
+    /* Copy all already arrived middle fragments to the data buffer */
+    ucs_queue_for_each_extract(m_rdesc, &ep_ext->am.mid_rdesc_q, am_mid_queue,
```

Comment:
is it possible that mid_rdesc_q would contain mixed fragments from different messages? in this case for_each_extract will not work since it stops the iteration when there is a mismatch and does not check others